gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -272,503 +307,306 @@ class GTSAM_EXPORT NoiseModelFactor: public NonlinearFactor {
+#define ALIAS_TRUE_X(NAME, N)                                                 \
+  template <class... VALUES>                                                  \
+  struct Alias##NAME##_<true, VALUES...> {                                    \
+    using NAME = typename std::tuple_element<N, std::tuple<VALUES...>>::type; \
+  };
+// Finally, alias a convenience struct that chooses the right version.
+#define ALIAS_X(NAME, N, CONDITION)     \
+  ALIAS_FALSE_X(NAME)        \
+  ALIAS_TRUE_X(NAME, N)      \
+  template <class... VALUES> \
+  using Alias##NAME = Alias##NAME##_<(CONDITION), VALUES...>;
+
+ALIAS_X(X, 0, 1 == sizeof...(VALUES));
+ALIAS_X(X1, 0, 0 < sizeof...(VALUES));
+ALIAS_X(X2, 1, 1 < sizeof...(VALUES));
+ALIAS_X(X3, 2, 2 < sizeof...(VALUES));
+ALIAS_X(X4, 3, 3 < sizeof...(VALUES));
+ALIAS_X(X5, 4, 4 < sizeof...(VALUES));
+ALIAS_X(X6, 5, 5 < sizeof...(VALUES));
+#undef ALIAS_FALSE_X
+#undef ALIAS_TRUE_X
+#undef ALIAS_X
+
+}  // namespace detail
 
+/* ************************************************************************* */
 /**
- * A convenient base class for creating your own NoiseModelFactor with 1
- * variable.  To derive from this class, implement evaluateError().
+ * A convenient base class for creating your own NoiseModelFactor
+ * with n variables.  To derive from this class, implement evaluateError().
  *
- * Templated on a values structure type. The values structures are typically
- * more general than just vectors, e.g., Rot3 or Pose3,
- * which are objects in non-linear manifolds (Lie groups).
+ * For example, a 2-way factor could be implemented like so:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * class MyFactor : public NoiseModelFactorN<double, double> {
+ *  public:
+ *   using Base = NoiseModelFactorN<double, double>;
+ *
+ *   MyFactor(Key key1, Key key2, const SharedNoiseModel& noiseModel)
+ *       : Base(noiseModel, key1, key2) {}
+ *
+ *   Vector evaluateError(
+ *       const double& x1, const double& x2,
+ *       boost::optional<Matrix&> H1 = boost::none,
+ *       boost::optional<Matrix&> H2 = boost::none) const override {
+ *     if (H1) *H1 = (Matrix(1, 1) << 1.0).finished();
+ *     if (H2) *H2 = (Matrix(1, 1) << 2.0).finished();
+ *     return (Vector(1) << x1 + 2 * x2).finished();
+ *   }
+ * };
+ * ~~~~~~~~~~~~~~~~~~~~
+ *
+ * These factors are templated on a values structure type. The values structures
+ * are typically more general than just vectors, e.g., Rot3 or Pose3, which are
+ * objects in non-linear manifolds (Lie groups).
  */
-template<class VALUE>
-class NoiseModelFactor1: public NoiseModelFactor {
-
-public:
-
-  // typedefs for value types pulled from keys
-  typedef VALUE X;
-
-protected:
-
-  typedef NoiseModelFactor Base;
-  typedef NoiseModelFactor1<VALUE> This;
-
-public:
+template <class... VALUES>
+class NoiseModelFactorN
+    : public NoiseModelFactor,
+      public detail::AliasX<VALUES...>,   // using X = VALUE1
```

Comment:
I strongly dislike this :-) clutters up a *key* class in GTSAM. All to have aliases X1 etc? Do we even use them anywhere?

Reply:
Another option is to just typedef them to void if they shouldn't exist.  This makes the code cleaner at the expense of less stringent compile-time checks.  For example, something like:

```c++
using X2 = std::conditional< (N > 2), VALUES[2], void >
```

but then someone might get confused if they do e.g. `NoiseModelFactorN<double, double>::X6 myVariable = 3.2` and then the compiler will give some confusing message (as opposed to just "X2 does not exist").




> Do we even use them anywhere?

I don't think it's common, but does occur in some places, e.g.
https://github.com/borglab/gtsam/blob/7f80c906c4cadc32e71acbe5209f2dba90ace779/tests/testExtendedKalmanFilter.cpp#L192-L194
or, strangely, `BoundingConstraint` redefines it:
https://github.com/borglab/gtsam/blob/7f80c906c4cadc32e71acbe5209f2dba90ace779/gtsam/slam/BoundingConstraint.h#L99-L102

But also, backwards compatibility?  What if users use it?

Reply:
@dellaert please advise on recommended course of action.  Options:
1) keep current
2) typedef to `void` if the type should not apply to this set of templates (cleaner using std::conditional).  e.g. `using X1 = std::conditional< (1 <= N), VALUES[0], void >`
3) delete X1, X2, ...  since they're rarely used (backwards compatibility violation)
4) change `#define NoiseModelFactor4 NoiseModelFactorN` to `struct NoiseModelFactor4 : public NoiseModelFactorN` and move these custom typedefs into there
5) other